apps/src/templates/instructions/InstructionsWithWorkspace.jsx

```diff
@@ -0,0 +1,183 @@
+    var maxHeight = instructionsContentHeight + HEADER_HEIGHT + RESIZER_HEIGHT;
+    if (maxHeight < topPaneHeight) {
+      topPaneHeight = maxHeight;
+    }
+
+    if (topSpaceAvailable < topPaneHeight) {
+      // if we'll still be at least 150px, just make our topPaneHeight smaller
+      if (topSpaceAvailable > INSTRUCTIONS_RESERVE) {
+        topPaneHeight = topSpaceAvailable;
+      } else {
+        topPaneHeight = Math.round(totalHeight / 3);
+      }
+      maxHeight = topPaneHeight;
+    } else if (topSpaceAvailable < maxHeight) {
+      maxHeight = topSpaceAvailable;
+    }
+    this.props.setInstructionsHeight(Math.min(topPaneHeight, maxHeight));
+    this.props.setInstructionsMaxHeight(maxHeight);
+  },
+
+  /**
+   * @returns {number} How much vertical space is consumed by the TopInstructions
+   */
+  topPaneHeight: function () {
+    // instructionsHeight represents the height of the TopInstructions if displayed
+    // and not collapsed
+    var height = this.props.instructionsHeight;
+
+    // If collapsed, we only use display instructions header
+    if (this.props.instructionsCollapsed) {
+      height = HEADER_HEIGHT;
+    }
+
+    // Or we may not display the instructions pane at all
+    if (!this.props.showInstructions) {
+      height = 0;
+    }
+
+    return height;
+  },
+
+  componentDidMount: function () {
+    if (this.props.showInstructions) {
+      this.adjustTopPaneHeight();
+
+      window.addEventListener('resize', this.onResize);
+    }
+  },
+
+  componentWillUnmount: function () {
+    if (this.props.showInstructions) {
+      window.removeEventListener("resize", this.onResize);
+    }
+  },
+
+  render: function () {
+    var topPaneHeight = this.topPaneHeight();
+
+    return (
+      <span>
+        {this.props.showInstructions && <TopInstructions
+            ref="topInstructions"
+            height={topPaneHeight}
+            onLoadImage={this.adjustTopPaneHeight}/>
+        }
+        <CodeWorkspaceContainer
+            ref="codeWorkspace"
+            topMargin={topPaneHeight}
+            noVisualization={false}
+            isRtl={false}
+            generateCodeWorkspaceHtml={this.props.generateCodeWorkspaceHtml}
+            onSizeChange={utils.fireResizeEvent}/>
+      </span>
+    );
+  }
+});
+
+module.exports = connect(function propsFromStore(state) {
+  return {
+    showInstructions: state.level.instructionsInTopPane && !!state.level.instructionsMarkdown,
+    instructionsCollapsed: state.instructions.collapsed || !state.level.instructionsInTopPane,
+    instructionsHeight: state.instructions.height,
+    instructionsMaxHeight: state.instructions.maxHeight,
+  };
+}, function propsFromDispatch(dispatch) {
+  return {
+    setInstructionsHeight: function (height) {
+      dispatch(instructions.setInstructionsHeight(height));
+    },
+    setInstructionsMaxHeight: function (maxHeight) {
+      dispatch(instructions.setInstructionsMaxHeight(maxHeight));
```

Comment:
These two props from dispatch aren't documented in the propTypes for this component.